envelope.go

```diff
@@ -42,69 +42,105 @@ var errSpecialEnvelope = errorf(
 type envelope struct {
 	Data  *bytes.Buffer
 	Flags uint8
+
+	offset int
 }
 
 func (e *envelope) IsSet(flag uint8) bool {
 	return e.Flags&flag == flag
 }
 
+// Read implements io.Reader.
+func (e *envelope) Read(data []byte) (readN int, err error) {
+	if e.offset < 5 {
+		prefix := makeEnvelopePrefix(e.Flags, e.Data.Len())
+		readN = copy(data, prefix[e.offset:])
+		e.offset += readN
+		if e.offset < 5 {
+			return readN, nil
+		}
+		data = data[readN:]
+	}
+	n := copy(data, e.Data.Bytes()[e.offset-5:])
+	e.offset += n
+	readN += n
+	if readN == 0 && e.offset == e.Data.Len()+5 {
+		err = io.EOF
+	}
+	return readN, err
+}
+
+// WriteTo implements io.WriterTo.
+func (e *envelope) WriteTo(dst io.Writer) (wroteN int64, err error) {
+	if e.offset < 5 {
+		prefix := makeEnvelopePrefix(e.Flags, e.Data.Len())
+		prefixN, err := dst.Write(prefix[e.offset:])
+		e.offset += prefixN
+		wroteN += int64(prefixN)
+		if e.offset < 5 {
+			return wroteN, err
+		}
+	}
+	n, err := dst.Write(e.Data.Bytes()[e.offset-5:])
+	e.offset += n
+	wroteN += int64(n)
+	return wroteN, err
+}
+
 type envelopeWriter struct {
-	writer           io.Writer
 	codec            Codec
 	compressMinBytes int
 	compressionPool  *compressionPool
 	bufferPool       *bufferPool
 	sendMaxBytes     int
 }
 
-func (w *envelopeWriter) Marshal(message any) *Error {
+func (w *envelopeWriter) Marshal(dst io.Writer, message any) *Error {
 	if message == nil {
-		if _, err := w.writer.Write(nil); err != nil {
+		if _, err := dst.Write(nil); err != nil {
 			if connectErr, ok := asError(err); ok {
 				return connectErr
 			}
 			return NewError(CodeUnknown, err)
 		}
 		return nil
 	}
-	if appender, ok := w.codec.(marshalAppender); ok {
-		return w.marshalAppend(message, appender)
+	buffer, err := w.marshal(message)
+	if err != nil {
+		return err
 	}
-	return w.marshal(message)
+	defer w.bufferPool.Put(buffer)
+	return w.Write(dst, &envelope{Data: buffer, Flags: 0})
 }
 
 // Write writes the enveloped message, compressing as necessary. It doesn't
 // retain any references to the supplied envelope or its underlying data.
-func (w *envelopeWriter) Write(env *envelope) *Error {
-	if env.IsSet(flagEnvelopeCompressed) ||
-		w.compressionPool == nil ||
-		env.Data.Len() < w.compressMinBytes {
-		if w.sendMaxBytes > 0 && env.Data.Len() > w.sendMaxBytes {
-			return errorf(CodeResourceExhausted, "message size %d exceeds sendMaxBytes %d", env.Data.Len(), w.sendMaxBytes)
+func (w *envelopeWriter) Write(dst io.Writer, env *envelope) *Error {
+	if !env.IsSet(flagEnvelopeCompressed) &&
+		w.compressionPool != nil &&
+		env.Data.Len() > w.compressMinBytes {
+		if err := w.compress(env.Data); err != nil {
```

Comment:
Behaviour change on `Write`, now mutates the `*env` to compress, if needed. Nothing currently depended on the `*env` being unchanged.